### PR TITLE
prov/efa: Use long CTS protocol if long read protocol fails because of memory registration limits

### PR DESCRIPTION
**Pull request by sunkuamzn**
The long read protocol can fail because of MR registration limits on the hardware. This PR has changes to switch to long CTS protocol when that happens.

### CODE REVIEW

**Comment by sunkuamzn** [comment]
@shefty could you review https://github.com/ofiwg/libfabric/pull/9432/commits/e13223a051939ee7c35d8439c13b6b7b5cdbba42?

**Comment by sunkuamzn** [comment]
bot:aws:retest

File: include/rdma/fi_errno.h [hunk 1]
```diff
@@ -194,6 +194,7 @@ enum {
 	FI_ENOAV	 = 267, /* Missing or unavailable address vector */
 	FI_EOVERRUN	 = 268, /* Queue has been overrun */
 	FI_ENORX	 = 269, /* Receiver not ready, no receive buffers available */
+	FI_ENOMR	 = 270, /* No more memory registrations available */
```
Comment (shefty) [inline include/rdma/fi_errno.h:197]: Looks good.  Should the fi_mr man page be updated to reference this error code?

Comment (sunkuamzn) [reply include/rdma/fi_errno.h:197]: FI_ENOMR is not returned to the application. It's only used internally within EFA provider. So I don't think we need to update the man page.

**Comment by sunkuamzn** [comment]
bot:aws:retest

**Comment by j-xiong** [comment]
@ofiwg/aws-efa-team Is it ready to merge?

**Comment by sunkuamzn** [comment]
Runting read fallback and doc update in https://github.com/ofiwg/libfabric/pull/9493

File: prov/efa/src/rdm/efa_rdm_msg.c [hunk 2]
```diff
@@ -158,7 +159,13 @@ ssize_t efa_rdm_msg_post_rtm(struct efa_rdm_ep *ep, struct efa_rdm_ope *txe, int
 	if (!efa_rdm_pkt_type_is_supported_by_peer(rtm_type, peer))
 		return -FI_EOPNOTSUPP;
 
-	return efa_rdm_ope_post_send(txe, rtm_type);
+	err = efa_rdm_ope_post_send(txe, rtm_type);
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_msg.c:162]: so `efa_rdm_ope_post_send_handle_error` currently handles the sender side registration failure? Better to have a comment here.

File: prov/efa/src/rdm/efa_rdm_rxe_map.c [hunk 3]
```diff
@@ -106,19 +106,21 @@ void efa_rdm_rxe_map_insert(struct efa_rdm_rxe_map *rxe_map,
...
  * @param[in,out]	rxe_map		RX entry map
- * @param[in]		pkt_entry	received RTM packet
+ * @param[in]		msg_id		message ID
+ * @param[in]		addr		peer address
  * @param[in]		rxe		RX entry
  */
 void efa_rdm_rxe_map_remove(struct efa_rdm_rxe_map *rxe_map,
-			    struct efa_rdm_pke *pkt_entry,
+			    uint64_t msg_id,
+				fi_addr_t addr,
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_rxe_map.c:115]: indent

File: prov/efa/src/rdm/efa_rdm_pke_rtm.c
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_rtm.c:1]: Better to document why this change is required in the commit message. I guess it's needed by the subsequent commits

File: prov/efa/src/rdm/efa_rdm_ope.c [hunk 4]
```diff
@@ -1575,15 +1576,23 @@ int efa_rdm_ope_post_remote_read_or_queue(struct efa_rdm_ope *ope)
...
+		dlist_insert_tail(&ope->queued_read_entry,
+				  &ope->ep->ope_queued_read_list);
 		ope->internal_flags |= EFA_RDM_OPE_QUEUED_READ;
 		err = 0;
-	} else if(err) {
-		EFA_WARN(FI_LOG_CQ,
-			"RDMA post read failed. errno=%d.\n", err);
+		break;
+	case -FI_ENOMR:
+		/* We want to fallback to emulated read, so just return FI_ENOMR without printing warning
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_ope.c:1587]: emulated read or Long CTS?

File: prov/efa/src/rdm/efa_rdm_pke_nonreq.c [hunk 5]
```diff
@@ -643,6 +662,42 @@ void efa_rdm_pke_handle_eor_recv(struct efa_rdm_pke *pkt_entry)
...
+	struct efa_rdm_ope *txe;
+	struct efa_rdm_peer *peer;
+
+	peer = efa_rdm_ep_get_peer(pkt_entry->ep, pkt_entry->addr);
+	assert(peer);
+	peer->num_read_msg_in_flight -= 1;
+
+	nack_hdr = (struct efa_rdm_read_nack_hdr *) pkt_entry->wiredata;
+
+	/* pre-post buf used here, so can NOT track back to txe with x_entry */
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_nonreq.c:680]: Not quite follow the comment here. the `pkt_entry` is from receiver, how can we use `pkt_entry->x_entry` as txe?

**Review by shijin-aws — COMMENTED** [review]
overall looks good

File: prov/efa/src/rdm/efa_rdm_pke_nonreq.c (hunk 5 above)
Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_pke_nonreq.c:680]: I copied this from the EOR packet processing code. I think it means what you're saying. I removed the comment.

File: prov/efa/src/rdm/efa_rdm_ope.c (hunk 4 above)
Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_ope.c:1587]: Long CTS. Fixed.

File: prov/efa/src/rdm/efa_rdm_pke_rtm.c
Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_pke_rtm.c:1]: Fixed

File: prov/efa/src/rdm/efa_rdm_rxe_map.c (hunk 3 above)
Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_rxe_map.c:115]: Fixed

File: prov/efa/src/rdm/efa_rdm_msg.c (hunk 2 above)
Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_msg.c:162]: Yes that's correct. What exactly do you want as the comment? I have a comment inside the `efa_rdm_ope_post_send_handle_error` function

Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_msg.c:162]: that is fine for now

**Review by shijin-aws — APPROVED** [review]

**Comment by j-xiong** [comment]
@shijin-aws Are you ok with getting this merged or still need comment from @wzamazon?

File: prov/efa/src/rdm/efa_rdm_protocol.h [hunk 6]
```diff
@@ -38,7 +38,8 @@ struct efa_ep_addr {
 #define EFA_RDM_EXTRA_REQUEST_CONNID_HEADER		BIT_ULL(3)
 #define EFA_RDM_EXTRA_FEATURE_RUNT			BIT_ULL(4)
 #define EFA_RDM_EXTRA_FEATURE_RDMA_WRITE		BIT_ULL(5)
-#define EFA_RDM_NUM_EXTRA_FEATURE_OR_REQUEST		6
+#define EFA_RDM_EXTRA_FEATURE_READ_NACK		BIT_ULL(6)
```
Comment (wzamazon) [inline prov/efa/src/rdm/efa_rdm_protocol.h:41]: indentation off.

File: prov/efa/src/rdm/efa_rdm_msg.c [hunk 7]
```diff
@@ -158,7 +159,13 @@ ssize_t efa_rdm_msg_post_rtm(struct efa_rdm_ep *ep, struct efa_rdm_ope *txe, int
...
+
+out:
+	if (OFI_UNLIKELY(err))
+		return efa_rdm_ope_post_send_handle_error(txe, rtm_type, err);
```
Comment (wzamazon) [inline prov/efa/src/rdm/efa_rdm_msg.c:166]: The function name is not clear. should be something like `post_send_fallback`.

File: prov/efa/src/rdm/efa_rdm_ope.c [hunk 8]
```diff
@@ -1729,6 +1729,35 @@ ssize_t efa_rdm_ope_post_send(struct efa_rdm_ope *ope, int pkt_type)
 	return 0;
 }
 
+ssize_t efa_rdm_ope_post_send_handle_error(struct efa_rdm_ope *ope,
```
Comment (wzamazon) [inline prov/efa/src/rdm/efa_rdm_ope.c:1741]: add a doxgen function header to explain what this function does.

Comment (wzamazon) [reply prov/efa/src/rdm/efa_rdm_ope.c:1741]: Would not it makes more sense for `efa_rdm_post_send` to call this function directly upon failure?

Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_ope.c:1741]: Sure, I can change it

Comment (sunkuamzn) [reply prov/efa/src/rdm/efa_rdm_ope.c]: Fixed

**Comment by j-xiong** [comment]
Replaced by #9493

**Pull request closed without merging**